django/db/migrations/state.py

```diff
@@ -107,6 +112,176 @@ def remove_model(self, app_label, model_name):
             # the cache automatically (#24513)
             self.apps.clear_cache()
 
+    def rename_model(self, app_label, operation):
+        # Add a new model.
+        renamed_model = self.models[app_label, operation.old_name_lower].clone()
+        renamed_model.name = operation.new_name
+        self.models[app_label, operation.new_name_lower] = renamed_model
+        # Repoint all fields pointing to the old model to the new one.
+        old_model_tuple = (app_label, operation.old_name_lower)
+        new_remote_model = '%s.%s' % (app_label, operation.new_name)
+        to_reload = set()
+        for model_state, name, field, reference in get_references(self, old_model_tuple):
+            changed_field = None
+            if reference.to:
+                changed_field = field.clone()
+                changed_field.remote_field.model = new_remote_model
+            if reference.through:
+                if changed_field is None:
+                    changed_field = field.clone()
+                changed_field.remote_field.through = new_remote_model
+            if changed_field:
+                model_state.fields[name] = changed_field
+                to_reload.add((model_state.app_label, model_state.name_lower))
+        # Reload models related to old model before removing the old model.
+        self.reload_models(to_reload, delay=True)
+        # Remove the old model.
+        self.remove_model(app_label, operation.old_name_lower)
+        self.reload_model(app_label, operation.new_name_lower, delay=True)
```

Comment:
As an example. The method signature should be

```python
    def rename_model(self, app_label, old_model_name, new_model_name):
        ...
```

And be called from `RenameModel.state_forwards` as `state.rename_model(app_label, self.old_name_lower, self.new_name_lower)` instead of passing the `Operation` instance along.